File: backend/app/schemas/investigation.py
from datetime import datetime
from typing import Generic
from typing import TypeVar

from pydantic import BaseModel
from pydantic import ConfigDict
from pydantic import Field

from backend.app.models.investigation import InvestigationStatus
from backend.app.models.investigation import InvestigationType
from backend.app.models.investigation import ModuleResultStatus
from backend.app.models.investigation import RiskLevel


# ==========================================================
# Create Investigation
# ==========================================================

class InvestigationCreate(BaseModel):

    investigation_type: InvestigationType

    target: str = Field(
        min_length=1,
        max_length=500,
    )


# ==========================================================
# Investigation Result
# ==========================================================

class InvestigationResultResponse(BaseModel):

    model_config = ConfigDict(
        from_attributes=True,
    )

    id: str

    source: str

    status: ModuleResultStatus

    data: dict | None = None

    latency_ms: int | None = None

    error_message: str | None = None

    created_at: datetime


# ==========================================================
# Investigation Response
# ==========================================================

class InvestigationResponse(BaseModel):

    model_config = ConfigDict(
        from_attributes=True,
    )

    id: str

    investigation_type: InvestigationType

    target: str

    status: InvestigationStatus

    risk_score: float | None = None

    risk_level: RiskLevel | None = None

    summary: str | None = None

    started_at: datetime | None = None

    completed_at: datetime | None = None

    error_message: str | None = None

    created_at: datetime

    updated_at: datetime

    results: list[InvestigationResultResponse] = []


# ==========================================================
# Investigation Summary (list view — no nested results)
# ==========================================================

class InvestigationSummary(BaseModel):

    model_config = ConfigDict(
        from_attributes=True,
    )

    id: str

    investigation_type: InvestigationType

    target: str

    status: InvestigationStatus

    risk_level: RiskLevel | None = None

    created_at: datetime


# ==========================================================
# Generic Paginated Response
# ==========================================================

T = TypeVar("T")


class PaginatedResponse(BaseModel, Generic[T]):

    items: list[T]

    total: int

    page: int

    page_size: int

    total_pages: int

File: frontend/src/types/investigation.ts
export type InvestigationType =
  | "username"
  | "email"
  | "domain"
  | "ip_address"
  | "dns"
  | "url"
  | "phone"
  | "metadata"
  | "reverse_image"
  | "file"
  | "breach"
  | "threat_intelligence"
  | "social_media"
  | "risk_assessment"
  | "malware";

export type InvestigationStatus =
  | "queued"
  | "running"
  | "completed"
  | "failed"
  | "partial";

export type RiskLevel = "low" | "medium" | "high" | "critical";

export type ModuleResultStatus =
  | "success"
  | "found"
  | "not_found"
  | "partial"
  | "unable_to_verify"
  | "no_data"
  | "rate_limited"
  | "failed"
  | "skipped";

export interface InvestigationResult {
  id: string;
  source: string;
  status: ModuleResultStatus;
  data: Record<string, unknown> | null;
  latency_ms: number | null;
  error_message: string | null;
  created_at: string;
}

export interface Investigation {
  id: string;
  investigation_type: InvestigationType;
  target: string;
  status: InvestigationStatus;
  risk_score: number | null;
  risk_level: RiskLevel | null;
  summary: string | null;
  started_at: string | null;
  completed_at: string | null;
  error_message: string | null;
  created_at: string;
  updated_at: string;
  results: InvestigationResult[];
}

export interface InvestigationSummary {
  id: string;
  investigation_type: InvestigationType;
  target: string;
  status: InvestigationStatus;
  risk_level: RiskLevel | null;
  created_at: string;
}

export interface PaginatedResponse<T> {
  items: T[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}
